pkg/util/mount/mount_linux.go

```diff
@@ -31,28 +32,80 @@ import (
 	"github.com/golang/glog"
 )
 
-const FlagBind = syscall.MS_BIND
-const FlagReadOnly = syscall.MS_RDONLY
+const (
+	// How many times to retry for a consistent read of /proc/mounts.
+	maxListTries = 3
+	// Number of fields per line in "/proc/mounts", as per the fstab man page.
+	expectedNumFieldsPerLine = 6
+)
 
 // Mounter implements mount.Interface for linux platform.
 type Mounter struct{}
 
-// Mount wraps syscall.Mount()
-func (mounter *Mounter) Mount(source string, target string, fstype string, flags uintptr, data string) error {
-	glog.V(5).Infof("Mounting %s %s %s %d %s", source, target, fstype, flags, data)
-	return syscall.Mount(source, target, fstype, flags, data)
+// Mount mounts source to target as fstype with given options. 'source' and 'fstype' must
+// be an emtpy string in case it's not required, e.g. for remount, or for auto filesystem
+// type, where kernel handles fs type for you. The mount 'options' currently come from
+// mount(8), if no more option is required, call Mount with an empty string list or nil.
+func (mounter *Mounter) Mount(source string, target string, fstype string, options []string) error {
+	glog.V(5).Infof("Mounting %s %s %s %v", source, target, fstype, options)
+	// Build mount command as follows:
+	//   mount [-t $fstype] [-o $options] [$source] $target
+	mountArgs := []string{}
+	if len(fstype) > 0 {
+		mountArgs = append(mountArgs, "-t", fstype)
+	}
+	if options != nil {
+		// If options is an empty list, this is a no-op.
+		mountArgs = append(mountArgs, options...)
+	}
+	if len(source) > 0 {
+		mountArgs = append(mountArgs, source, target)
+	} else {
+		mountArgs = append(mountArgs, target)
+	}
+	command := exec.Command("mount", mountArgs...)
+	output, err := command.CombinedOutput()
+	if err != nil {
+		glog.Errorf("Mount failed: %v\nMounting arguments: %s %s %s %v\nOutput: %s\n",
+			err, source, target, fstype, options, string(output))
+		return err
+	}
+	// Do a remount if caller asks for readonly bind - the previous mount call
+	// will NOT respect readonly option and will mount target as read/write.
+	if isReadOnlyBind(options) {
+		glog.V(5).Infof("Remounting %s as readonly bind", target)
+		command = exec.Command("mount", "-o", "remount,ro", target)
```

Comment:
If it was -o bind above and the system does not have a writable /etc/mtab, this will not act correctly. It would be most reliable to make sure if you have bind above you also have it here.

As to @rootfs maybe it owuld be wise to ban "remount" above and just make sure it is handled correctly here, until such time we have a need to support "remount" above....